src/platform/packages/private/kbn-grid-layout/grid/grid_panel/use_panel_grid_data.tsx

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { useEffect, useMemo, useRef } from 'react';
+import { BehaviorSubject, distinctUntilChanged, map } from 'rxjs';
+
+import type { GridSectionData } from '../grid_section';
+import type { OrderedLayout } from '../types';
+import { useGridLayoutContext } from '../use_grid_layout_context';
+import { isGridDataEqual } from '../utils/equality_checks';
+import type { GridPanelData } from './types';
+
+export const useGridPanelState = ({
+  panelId,
+}: {
+  panelId: string;
+}): BehaviorSubject<(GridPanelData & { sectionId: string }) | undefined> => {
+  const { gridLayoutStateManager } = useGridLayoutContext();
+  const cleanupCallback = useRef<null | (() => void)>();
+
+  const panel$ = useMemo(() => {
+    const panelSubject = new BehaviorSubject(
+      getPanelState(gridLayoutStateManager.gridLayout$.getValue(), panelId)
+    );
+
+    const subscription = gridLayoutStateManager.gridLayout$
+      .pipe(
+        map((layout) => getPanelState(layout, panelId)),
+        distinctUntilChanged(
+          (panelA, panelB) =>
+            isGridDataEqual(panelA, panelB) && panelA!.sectionId === panelB!.sectionId
+        )
+      )
+      .subscribe((panel) => {
+        panelSubject.next(panel as GridPanelData & { sectionId: string }); // undefined panels were filtered out
+      });
+
+    cleanupCallback.current = () => {
+      subscription.unsubscribe();
+    };
+
+    return panelSubject;
+  }, [gridLayoutStateManager.gridLayout$, panelId]);
+
+  useEffect(() => {
+    return () => {
+      if (cleanupCallback.current) cleanupCallback.current();
+    };
+  }, []);
+
+  return panel$;
+};
+
+const getPanelState = (
```

Comment:
actually, I already put 2 nitpicky comments to this function, but I think it's better to write this function a bit different:

```
const getPanelState = (
  layout: OrderedLayout,
  panelId: string
): (GridPanelData & { sectionId: string }) | undefined => {
  for (const section of Object.values(layout)) {
    const panel = section.panels[panelId];
    if (panel) {
      return { ...panel, sectionId: section.id };
    }
  }
  return undefined;
};
```

Why it's better:
1. stops searching when the panel is found
2. doesn't allocate memory for all panels only to return one we look for

Reply:
Nice!! Very clean. I wrote this very quickly and then never revisited it, oops 🙈  Done in 657c0f1ac13da5715e848330e2312409d7ef997